src/codeu/chat/client/ClientMessage.java

```diff
@@ -104,40 +98,76 @@ public void addMessage(Uuid author, Uuid conversation, String body) {
 
     if (message == null) {
       System.out.format("Error: message not created - %s.\n",
-          (validInputs) ? "server error" : "bad input value");
+              (validInputs) ? "server error" : "bad input value");
     } else {
       LOG.info("New message:, Author= %s UUID= %s", author, message.id);
       current = message;
+
     }
+
     updateMessages(false);
   }
 
-  // Delete message, (m-del command?).TODO
+  // Delete message, removes last message
+  // m-del-last command
   public void deleteMessage() {
-    if(current != null) {
-      deleteMessage(current);
+    if (currentMessageCount() == 0) {
+      System.out.println("Current Conversation has no messages.");
+
     } else {
-    } 
+      if (controller.deleteMessage(conversationHead.lastMessage, conversationHead.id)) {
+        resetCurrent(currentMessageCount() == 1);
+
+      } else {
+
+      }
+    }
   }
 
-  public void deleteMessage(Message msg) { //TODO: Use an ordered hash map for linear time. https://github.com/google/guava
-    if(conversationContents.contains(msg)) {
-      conversationContents.remove(msg);
+  // Delete message, removes message corresponding to given index, (m-delete <index> command)
+  // calls helper method
+  public void deleteMessage(String index) { //TODO: Use an ordered hash map for linear time. https://github.com/google/guava
+    int msgIndex = Integer.valueOf(index);
+    if (msgIndex == (currentMessageCount() - 1)) {
+      // Message to be deleted is the last message in the conversation,
+      // so m-del-last command is called
+      deleteMessage();
+
     } else {
-      System.out.println("Error: message not found.");
+      if (msgIndex < currentMessageCount()) {
+        Message msg = conversationContents.get(msgIndex);
+        deleteMessage(msg);
+        updateMessages(true);
+
+      } else {
 
+        LOG.error(" Error: message not found, please enter a valid index.");
+      }
     }
   }
 
 
-  // Delete all messages.TODO
+  // Delete message helper method
+  public void deleteMessage(Message msg) {
+    if (currentMessageCount() == 0) {
+      LOG.error("Error: Message could not be deleted, current Conversation has no messages");
+
+    } else if (currentMessageCount() == 1) {
+      deleteMessage();
+
+    } else {
+        controller.deleteMessage(msg.id, conversationHead.id);
+    }
```

Comment:
controller.deleteMessage(msg.id, conversationHead.id) does handle everything. The second if-statement was to make sure that deleteMessage(Message msg) would call the appropriate method if it were to be asked to delete the last message in a conversation. Since that method should only be called internally, I changed the privacy setting to private and got rid of that second branch.